pkg/execution/state/redis_state/backlog.go

```diff
@@ -0,0 +1,284 @@
+	ConcurrencyKeyValue *string `json:"ckv,omitempty"`
+
+	// ConcurrencyKeyUnhashedValue is the unhashed evaluated key (e.g. "customer1")
+	// This may be truncated for long values and may only be used for observability and debugging.
+	ConcurrencyKeyUnhashedValue *string `json:"ckuv,omitempty"`
+
+	// Set for backlogs containing start items only for a given throttle configuration
+
+	// ThrottleKey is the hashed evaluated throttle key (e.g. hash("customer1")) or function ID (e.g. hash(fnID))
+	ThrottleKey *string `json:"tk,omitempty"`
+
+	// ThrottleKeyRawValue is the unhashed evaluated throttle key (e.g. "customer1") or function ID.
+	// This may be truncated for long values and may only be used for observability and debugging.
+	ThrottleKeyRawValue *string `json:"tkv,omitempty"`
+}
+
+func (q *queue) ItemBacklogs(ctx context.Context, i osqueue.QueueItem) []QueueBacklog {
+	backlogs := make([]QueueBacklog, 0)
+
+	queueName := i.QueueName
+
+	// sanity check: both QueueNames should be set, but sometimes aren't
+	if queueName == nil && i.QueueName != nil {
+		queueName = i.QueueName
+		q.logger.Warn().Interface("item", i).Msg("backlogs encountered queue item with inconsistent custom queue name, should have both i.QueueName and i.Data.QueueName set")
+	}
+
+	// sanity check: queueName values must match
+	if i.Data.QueueName != nil && i.QueueName != nil && *i.Data.QueueName != *i.QueueName {
+		q.logger.Error().Interface("item", i).Msg("backlogs encountered queue item with inconsistent custom queue names, should have matching values for i.QueueName and i.Data.QueueName")
+	}
+
+	if queueName != nil {
+		// Simply use default backlog for system queues - there shouldn't be any concurrency or throttle keys involved.
+		backlogs = append(backlogs, QueueBacklog{
+			BacklogID: fmt.Sprintf("system:%s", *queueName),
+		})
+		return backlogs
+	}
+
+	// Enqueue start items to throttle backlog if throttle is configured
+	if i.Data.Throttle != nil && i.Data.Kind == osqueue.KindStart {
+		b := QueueBacklog{
+			BacklogID: fmt.Sprintf("throttle:%s:%s", i.FunctionID, i.Data.Throttle.Key),
+
+			// This is always specified, even if no key was configured in the function definition.
+			// In that case, the Throttle Key is the hashed function ID. See Schedule() for more details.
+			ThrottleKey: &i.Data.Throttle.Key,
+		}
+
+		if i.Data.Throttle.UnhashedThrottleKey != "" {
+			unhashedKey := i.Data.Throttle.UnhashedThrottleKey
+			// truncate - just in case
+			if len(unhashedKey) > 512 {
+				unhashedKey = unhashedKey[:512]
+			}
+			b.ThrottleKeyRawValue = &unhashedKey
+		}
+
+		backlogs = append(backlogs, b)
+	}
+
+	concurrencyKeys := i.Data.GetConcurrencyKeys()
+
+	// NOTE: This is an optimization that ensures we return *updated* concurrency keys
+	// for any recently published function configuration.  The embeddeed ckeys from the
+	// queue items above may be outdated.
+	if q.customConcurrencyLimitRefresher != nil {
```

Comment:
this whole section can probably just be another helper function.

main thing is we're retrieving the `concurrencyKeys`, how and when it needs to be refreshed probably can be encapsulated elsewhere.

Reply:
this could probably also return a map instead of a list so it's not quadratic.